pkg/store/gate.go

```diff
@@ -0,0 +1,64 @@
+package store
+
+import (
+	"context"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/extprom"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/gate"
+)
+
+// Gate wraps the Prometheus gate with extra metrics.
+type Gate struct {
+	g               *gate.Gate
+	inflightQueries prometheus.Gauge
+	gateTiming      prometheus.Histogram
+}
+
+// NewGate returns a new query gate.
+func NewGate(maxConcurrent int, reg *extprom.SubsystemRegisterer) *Gate {
+	g := &Gate{
+		g: gate.New(maxConcurrent),
+	}
+	g.inflightQueries = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name:      "queries_in_flight_total",
```

Comment:
prefix thanos? I think this is out of sync with the documentation

Reply:
`thanos_bucket_store_queries_in_flight` is in the documentation ... is this code only going to be used in the store-gateway or is it able to be used in all of the storeAPI endpoints?